### ISSUE TITLE
Determining the best width

### ISSUE THREAD

**Issue by SteveTheTechie** [ehynds/jquery-ui-multiselect-widget#727]
Reference issue #717, comments 2 and 8.

Here is the essence of how I automatically determine a "best width" that will accommodate all the select options...  Essentially a hidden clone of the menu is produced that is used for measuring.  The clone is set up specifically to allow the container div to assume the smallest width needed to accommodate the widest option.

Determining a single option height allows us to just multiply the number of options by the native select's size attribute to determine a menu height.

This is not perfect-there is room for improvement.  In particular, whether or not options can be wrapped should be set via a new option.  Wrapping affects the width and the appearance.

Also, the measuring can be a bit slow.   If we had a reliable way to quickly determine the widest option then we could just make a dummy menu clone with that single option for measuring which would be faster.  This is an approach that I not yet explored.

```
	// Added to allow automatic menu sizing,  8/17/2014, Steve James
	// Reference: http://stackoverflow.com/questions/1472303/jquery-get-width-of-element-when-not-visible-display-none
	// 		http://stackoverflow.com/questions/5903112/is-it-really-impossible-to-make-a-div-fit-its-size-to-its-content
	
	_measureMenu: function() {			
		// width: auto needed to reset width when dialogs 
                //    containing multi-select are re-opened.  
		// Display: inline-block needed to force width to best fit.
		
		// Make the hidden clone for measuring...
		var $clone = this.$menu.clone()
                                 .appendTo('body')
                                 .css({visibility:'hidden',width:'auto',display:'inline-block'})
                                 .find('ul')
                                 .css({display:'inline-block'})
                                 .end()
                                 .show();
				
		// Find outer height of a hidden menu item before we display menu.																											
		this.optionHeight = $clone.find('label').outerHeight();																																	
		
		// Find best width of hidden menu before we display it; 20 is for scroll bar		
		this.menuWidth = Math.max($clone.find('.ui-multiselect-header').width(),
                                         $clone.find('.ui-multiselect-checkboxes').width()) - 20;
		$clone.remove(); 
	},

```

**Comment by SteveTheTechie** [comment]
I am currently digging into this again after several years.  Getting reliable measurements of DOM elements can be a real challenge at times.   Some of the relevant understandings that I have gained over time and via recent research:

1.  Measurements can only be obtained from elements that take up space in the page layout.  Elements with `display:none` (hidden) do not occupy space and therefore heights and widths will be zero for those elements.  This is why if you look carefully, dimensions of the menu are obtained *after* the menu is opened (shown).  Thus, the point in time when you take measurements is a relevant and important consideration.  In the case of this widget, I have to wonder about how the time for animations may affect getting correct measurements.  Perhaps we should add a .then() somewhere (Deferreds are supported for jQuery animations).

2.  Hidden elements can be measured by displaying them or by setting visibility as 'hidden'.  Visibility='hidden' means that space is allocated in the layout but the elements are not shown.  (so you could see big blank spots in the layout)  One way to address this is to either move or clone them to the end of the body, out of view, sort of like what I do above.

3.  jQuery appears to do some "magic" in some cases to get and set dimensions of even hidden elements.  My understanding is that, in reality, jQuery actually will monkey with the DOM in some cases to force a situation where it can actually get a measurement.  The jQuery methods can cause reflow lags... these can be significant depending on what is being measured.

4.  However, even though jQuery has its ways of getting / setting measurements, they are not perfect, and they have changed with different jQuery versions.   This means that any thing that uses the jQuery measurement methods may have jQuery version dependencies and may get different results depending on which jQuery version is used.    As an example, take a look at the following:
[https://github.com/jquery/jquery/issues/3193](https://github.com/jquery/jquery/issues/3193)

**Comment by SteveTheTechie** [comment]
My thoughts on how to improve the widget re: measurements.

1.  Measure / set dimensions once only and cache the computed results.  If cached value is detected, either use it to set dimensions (in some cases) or avoid redundantly setting measurements.  Certain types of callbacks or events may invalidate the cache, which forces the values to be recalculated on the next open.  Additionally, the relevant measurement functions may take a `reset` parameter, which forces the cached value to be recalculated.

2.  The _setMenuHeight function iterates the heights of *all* menu items.   This is totally unnecessary, particularly with the size option I am adding to use the native select's size attribute for sizing the menu.  The key insight is that once the loop has determined that the summed height of the examined menu items exceeds either option height setting or if the number of examined items equals the native select's size attribute (if size option being used), the the loop can abort.  The only purpose of the loop is to determine what height is needed to contain "# items shown=size attribute" or if the actual height is less than the specified option height.    For large numbers of options, being able to able the vertical size summation loop is an optimization.

**Comment by SteveTheTechie** [comment]
_getOuterWidth & _getOuterHeight functions (no jQuery version dependencies)... notice that they can also handle hidden elements:

```
_getOuterWidth: function(el, withMargins) {
	if (!el) 
		return null;
	 
	var el_style = window.getComputedStyle ? getComputedStyle(el) : el.currentStyle;	// el.currentStyle is IE-compatible		 
	if (el_style.display === 'none' || parseInt(el_style.maxWidth, 10) == 0)
		return this._unHide4Dims(el, this._getOuterWidth, withMargins);
		 
	var width = el.offsetWidth || 0;
	if (!!withMargins)
		width += parseInt(el_style.marginLeft) + parseInt(el_style.marginRight);
	
	return width;		 
 },
 
 _getOuterHeight: function(el, withMargins) {
	if (!el) 
		return null;
	 
	var el_style = window.getComputedStyle ? getComputedStyle(el) : el.currentStyle;	// el.currentStyle is IE-compatible		 
	if (el_style.display === 'none' || parseInt(el_style.maxHeight, 10) == 0)
		return this._unHide4Dims(el, this._getOuterHeight, withMargins);
	 
	var height = el.offsetHeight || 0;
	if (!!withMargins)
		height += parseInt(el_style.marginTop) + parseInt(el_style.marginBottom);
	
	return height;		 
 },
 
 // Temporarily unhides hidden elements to allow dimensions to be obtained.
 // Reference: https://stackoverflow.com/questions/1473584/need-to-find-height-of-hidden-div-on-page-set-to-displaynone
_unHide4Dims: function(el, fxn, withMargins) {
	 if (!el) return null;
	 
	var el_style = window.getComputedStyle ? getComputedStyle(el) : el.currentStyle;	// el.currentStyle is IE-compatible
	var el_display = el_style.display;
	var el_position = el_style.position;
	var el_visibility = el_style.visibility;
	
	el.style.position = 'absolute';
	el.style.visibility = 'hidden';
	el.style.display = 'block';
	
	var dim = fxn.call(this, el, withMargins) 
	
	el.style.display = el_display;
	el.style.position = el_position;
	el.style.visibility = el_visibility;
	
	return dim;
}

```

**Comment by SteveTheTechie** [comment]
Measurement to-do's (some are done in my "pre-PR" code):
- [x] 1. Implement caching of dimensions and a way to invalidate the caching.
- [x] 2. Rework the _setMenuHeight function.
- [ ] 3. Confirm whether the measurement of _outer_ height of the native select for % calc should instead be an _inner_ height basis.
- [ ] 4. Implement a text wrapping option and a way to manage it.
- [ ] 5. Implement a minWidth: 'auto' option and a way to determine best fit width.  (maybe cloning method)

Your thoughts?

**Comment by mlh758** [comment]
I was going to say 1 might be a bit heavy if we can avoid it given the improvements of the others, but if you've already implemented it then fair enough.

2. I certainly agree, it would probably reduce the menu open time of that 5,000 item list in the speed test substantially since it would stop much sooner.

3. I wasn't even aware of the existence of `clientWidth`. I was using `outerWidth` because everything else in here was using it so it kept things consistent internally with the widget. It's also what people [seemed to expect](https://github.com/ehynds/jquery-ui-multiselect-widget/issues?utf8=%E2%9C%93&q=is%3Aissue+outerWidth). However, it would seem reasonable to size the percentage based on the actual element box rather than grabbing the outer boxes of margin and padding as well.

4. Useful to have, should be in a separate issue though.

5. What would be an expected algorithm for a sane minimum width in 90% of cases? Maybe let this parameter be a function instead so the developer can pass it in at their discretion?

**Comment by SteveTheTechie** [comment]
I did a test with your gist code wherein I wrapped the button in a 250px wide div with a 1px red border, then set the minWidth option of the multiselect to be '100%'.

When I do this with the current code, the right edge of the button obscures the red right border of the div.   

If I change from using outerWidth of the parent to using innerWidth in getMinWidth(), no overlap happens and the button fits perfectly.

Therefore, I conclude from this test that my assertion is valid that innerWidth should be used instead of outerWidth for % widths.

Before I had a hunch, now with this test it is proven, and I will make the change.

**Comment by mlh758** [comment]
I appreciate you doing the test first. Thanks for making the change.

**Comment by SteveTheTechie** [comment]
@mlh758 : I request that we collaborate on this a bit.  There are a number of subtle nuances to consider in setting / determining widths that factor into this. It is not that difficult... just a number of things to factor in.

1. I can kill text wrapping now for the menu... I essentially have an option for that in my WIP code and can see it working. 
2. There is a good bit of CSS that factors into the width--you cannot get around that.  For example, the jQuery UI ui-state-hover class adds 4px width to each menu item everytime you hover on an item. (2px for l/r border and 2px for l/r padding)  Thus, a 4px minimum "fudge factor" is needed to account for this.  The bolding of the hovered text also adds width, so some factor would have to be added for that, also.
3. The extreme test case that helps is setting minWidth = 0 in the options.  This is a nonsense value, but it does force the menu width to be set to whatever the minimum button width is.  This is a narrow width that helps see the effect of shutting off the text wrapping in the menu.
4. The minimum required width of the menu can really only be determined once the menu is opened, since widths of hidden elements cannot be reliably determined.  To get a proper "minimum" value, the text wrapping must be shut off.
5. The menuWidth option overrides whatever minWidth is determined and sets the menuWidth directly.  This kind of makes the minWidth option pointless and redundant in that case.  Also, I wonder if this could result in a menu that is a different size than the button... Isn't that stylistically bad?
6.  Here are some questions:
a. Should the button width and the menu width be aligned and equal?  (I tend to think so.)
b. Right now, the menuWidth option is not checked when setting the button width.  If the button and menu are supposed to be equal width, I would think you would check the menu width when setting the button width.
c. Why are two width options needed?  Since some string parsing of minWidth is already being done, couldn't we just have a menuWidth option, and have '>=150' mean 150px is a minWidth or '>=50%' is a minWidth?  Alternatively, a setting of 150 (number or string) would just be a fixed menu width.
d. When I look at the Harvest Chosen widget options [https://harvesthq.github.io/chosen/](https://harvesthq.github.io/chosen/), then allow setting a fixed % width.  I think that would be good also.
e. This really just ends up being a case of keeping the button width in sync with the menu width if the intention is to keep them equal.  You set the button with either a fixed menu width or the minimum button width needed, and if no fixed menu width, then you determine a best menu width on the first menu open (and cache it) and adjust the button width w/ that.

**Comment by mlh758** [comment]
Your first 5 points I'll have to think about a little bit, but I can answer a couple of the questions now:

a. I don't necessarily think so. The button only really needs to accommodate the header links and whatever message is displayed for checked items (3 of 6 selected for example). The menu can be much wider. Useful if a form doesn't have a lot of space, or you want to put several selects in a row.

b. This is because of the assumption in a, that a wider menu could be useful. I'm open to challenging this assumption, more modern widgets like Chosen and Select2 seem to keep the button and menu the same width.

c. This is a good idea that could help simplify the options. We could even leave a lot of the internal code the same and just process the width option first and setting the minWidth and width options appropriately.

d. This would pair well with c since a lot of the same logic can be used.

e. It seems like we would need to have the menu visible first, then adjust the button's width, then hide the menu rather than waiting until the first opening. Otherwise the button's width would change after the first time the widget was opened and that would be surprising. Unless I'm misunderstanding.

There seems to be two options for getting the measurements:
1. Clone the menu somewhere like you suggested earlier, take the measurements, and apply them to the widget. The downside is that this can be expensive.

2. Start with the menu open, then close it (unless autoOpen is true) once measurements are done. The downside here is that with long lists the menus would probably be briefly visible to the user before hiding.


It seems like it might be a good idea to drop the multitude of width options and go with a menu and button that are the same width at all times. The minWidth option, from this perspective, seems like a bandaid to deal with wide options on a narrow button whose width was determined independently of the menu. If the button's width is determined based on the menu a minimum width seems less necessary. You would just need a `width` option to override the calculations if you wanted consistently sized buttons in a form. This is where text wrapping or ellipses might come back into play.

**Comment by SteveTheTechie** [comment]
> a. I don't necessarily think so. The button only really needs to accommodate the header links and whatever message is displayed for checked items (3 of 6 selected for example). The menu can be much wider. Useful if a form doesn't have a lot of space, or you want to put several selects in a row.
> 
> b. This is because of the assumption in a, that a wider menu could be useful. I'm open to challenging this assumption, more modern widgets like Chosen and Select2 seem to keep the button and menu the same width.

I have to wonder if the "groupColumns" option maybe comes into play for this to allow sufficient menu width--I do not know if Chosen or Select2 have a comparable option.  This option was not in my original fork of a couple years ago, and I did not originally understand it.  I have since fiddled around with it a bit... Even though the layout was a bit jarring to look at first, I now think it is an innovative layout, and I can see where it could be useful.

> c. This is a good idea that could help simplify the options. We could even leave a lot of the internal code the same and just process the width option first and setting the minWidth and width options appropriately.
> 
> d. This would pair well with c since a lot of the same logic can be used.

Ok, I will move in the direction of a single, unified width option then.

> e. It seems like we would need to have the menu visible first, then adjust the button's width, then hide the menu rather than waiting until the first opening. Otherwise the button's width would change after the first time the widget was opened and that would be surprising. Unless I'm misunderstanding.

I think you got it.  However, I am not sure how obvious the button width change would be... it depends on how much of a change.  

Also, you have kind of sold me on not necessarily keeping the button width the same as the menu width, particularly for the sake of the groupColumns option.

Also, you are correct that doing a clone to check width is expensive.  That is why I continued to use fixed width options even after I had created an "auto" width option.  However, the auto width is useful for cases where options are added dynamically and the native select's parent element does not constraint width or in cases where I just was lazy (lol) and did not want to figure out a good width.

**Comment by SteveTheTechie** [comment]
The following looks interesting:  
[https://stackoverflow.com/questions/5429372/how-to-set-width-of-li-based-on-its-longest-contents-width](https://stackoverflow.com/questions/5429372/how-to-set-width-of-li-based-on-its-longest-contents-width)

However, I do think outerWidth() needs to be used as opposed to css('width')... this is in one of the comments on this page if you dig deep enough.

**Comment by mlh758** [comment]
> Meanwhile, block level elements fill the width of their container by default. What if we want the width of a block-level container to be determined by its content? Here's where intrinsic sizing comes in, currently supported by all browsers except IE & Edge.

Ah Microsoft...

One advantage we have is that we already have to process all the options in the select to build out the option list. Assuming we go down the route of not hiding the menu until after it is built, this would let us take a measure of each element and essentially keep a running `maxWidth` as we build the menu. Once it's done we can apply that to the button and menu, plus the fudge factor you mentioned earlier if necessary.

> Also, you have kind of sold me on not necessarily keeping the button width the same as the menu width, particularly for the sake of the groupColumns option.

You should have let me let you talk me out of it! If we want to keep the menuWidth option we could just apply that to the menu after determining the menu/button width via the algorithm.

**Comment by SteveTheTechie** [comment]
What about maybe having both buttonWidth and a menuWidth options?   If only one is set to a value (vs null), then the implication could be that both are set to the same as appropriate.   We could support % widths on the buttonWidth, but probably not on the menuWidth.  (since by design, it pops up and overlays content) 

If we support using '>=150' or similar for each, this would be clearer than a minWidth option which is kind vague about what width the minimum applies to.

You are correct that we could maybe do a width determination as we build the option list.  However, keep in mind that the content must be visible (in the visible layout) to get dimensions and that could factor in.  Another confounding challenge that I am aware of is that the parent elements of whatever you are measuring can have an impact on the determined width.   (e.g., if the parent element's width constrains the child element's width)  This is why if you measure a clone, it should be appended to the body so that the width is not constrained.

Let me do a bit more research and fiddling on this.  I know that this is something that has been handled by others... I just need to find the approach.

**Comment by mlh758** [comment]
> You are correct that we could maybe do a width determination as we build the option list. However, keep in mind that the content must be visible (in the visible layout) to get dimensions and that could factor in. 

Ah, right, and we don't actually append anything to the dom until they're all ready. Wouldn't do any good. A workaround might be simply taking the text length of each item and storing a reference to the longest element to take it's measure afterwards.

> If we support using '>=150' or similar for each, this would be clearer than a minWidth option which is kind vague about what width the minimum applies to.

True enough. Would it be any easier to start with matching widths, no less than, and then build up from there? This is already looking like a re-write of width behavior, it might be easier to start with automatic and then bolt on overrides.

**Comment by SteveTheTechie** [comment]
Here is the code that Select2 is using for autoWidth (it took me a bit to find it):

```
  AttachBody.prototype._resizeDropdown = function () {
    var css = {
      width: this.$container.outerWidth(false) + 'px'
    };

    if (this.options.get('dropdownAutoWidth')) {
      css.minWidth = css.width;
      css.position = 'relative';
      css.width = 'auto';
    }

    this.$dropdown.css(css);
  };

  AttachBody.prototype._showDropdown = function (decorated) {
    this.$dropdownContainer.appendTo(this.$dropdownParent);

    this._positionDropdown();
    this._resizeDropdown();
  };

  return AttachBody;
});

```

**Comment by SteveTheTechie** [comment]
Useful:
[https://github.com/select2/select2/pull/4471](https://github.com/select2/select2/pull/4471)

**Comment by SteveTheTechie** [comment]
Button 'auto width' (menu 'auto width' forthcoming)...

1.  Most useful for showing selection lists on a single line.  The button "grows" horizontally.
2.  Basically, set CSS min-width to be the width of the native select (as a reasonable starting width), and set CSS width to be 'auto'.  This causes the button to get wider as more options are selected.
3.  Text wrapping does not matter for this case since the button either gets wider to accommodate the text (width is auto) or the text is wrapped if the button width is fixed at some value.  It does not seem like it would make sense to use ellipses to indicated that non-wrapped text extends beyond the view.
4.  Considerations:
 - Unless a max-width is set to cause wrapping at some point (w/ enough list selections shown), the button will grow in width forever.   Therefore, an implicit max-width equal to window.innerWidth would be wise--this would cause the displayed list selections to wrap at that point.   
 - Perhaps 'auto' is not best for this case.  Perhaps setting a max width is better and using the native select as the min width.  That seems like it would be better for actual use.
 - Whether or not to synchronize the menu width is a consideration.

**Comment by SteveTheTechie** [comment]
Update on menu 'auto width' determination:

1.  Good references:  
- https://stackoverflow.com/questions/450903/how-to-make-div-not-larger-than-its-contents)  
- https://www.google.com/search?q=css+shrink+wrap+parent+and+child+divs&oq=css+shrink+wrap+parent+and+child+divs&aqs=chrome..69i57.11853j0j7&sourceid=chrome&ie=UTF-8
- https://stackoverflow.com/questions/26390119/make-parent-div-with-absolute-position-take-the-width-of-children-divs
- https://stackoverflow.com/questions/21998843/increase-width-on-overflow-to-adjust-scrollbars
2.  I do not think the clone method will be necessary.  
3.  I have experimented by simply temporarily setting `float:left` on the header and checkboxes to get the  "shrink-to-fit" (intrinsic) width for each. What it does not account for are scrollbar with and width needed for the hover highlight.   I have not (yet) been able to make the header and checkboxes shrink to fit *and* also have the outer menu container do the same.  Therefore, it is probably a case of using the widths of the header and checkboxes to set the "auto width" of the container.
4.  Scrollbar width can be determined from $.position.scrollbarWidth() or a polyfill. It only needs to be determined once for all instances. (Determine once and save.)  The need to add in scroll bar width is known by whether .css('overflow-y') comes back as auto, as this is set by the menu height determination.
5.  Select2 is having issues with width calcs, also.  (https://github.com/select2/select2/issues/4224)  They are moving to using getBoundingClientRect to avoid issues w/ floating point round off errors that are present when using the jQuery width functions.

Bottom Line: I think I have learned just about everything I need for this.  Trying to figure out the best combination to get the best results.

**Comment by SteveTheTechie** [comment]
New internal helper function that parses dimensional information into pixel return values:

```
 /**
  * Converts dimensions specified in options to pixel values.
  * Determines if specified value is a minimum, maximum or exact value.
  * The value can be a number or a string with px, pts, ems, or % units.
  * Number/Numeric string treated as pixel measurements
  *  - 30
  *  - '30'
  *  - '>30px'
  *  - '1.3em'
  *  - '20 pt'
  *  - '30%'
  * @returns {pixels, minimax} object containing pixels and -1/1/0 indicating min/max/exact.
  */
 _parse2pixels: function(dimText, elem) {
	if (typeof dimText !== 'string') 
		return {pixels: dimText, minimax: 0};
	
	var parts = dimText.match(/([<>])?=?\s*([.\d]+)\s*([emptsx%]*)/i);
	var minimax = parts[1];
	var value = parseFloat(parts[2]);
	var unit = parts[3];
	// console.log('minimax=', minimax, ' value=', value, ' unit=', unit);
	var pixels;
	switch (unit.toLowerCase()) {
		case 'pt': 
		case 'pts': 
			pixels =  4.0 / 3.0 * value;
			break;
		case 'em':
		case 'ems':
			var body = document.getElementsByTagName('body')[0];
			var bodyFontSize = (window.getComputedStyle ? getComputedStyle(body).fontSize : body.currentStyle.fontSize) || '16px';
			pixels = parseFloat(bodyFontSize) * value;
			break;
		case '%':
			if ( !!elem ) {
				pixels = $(elem).parent().innerWidth() * (value/100.0);
			}
			break;
		default:
			pixels = value;
	}
	// minimax:  -1 => minimum value, 1 => maximum value, 0 => exact value
	return {pixels: pixels, minimax: minimax == '>' ? -1 : minimax == '<' ? 1 : 0};
 },

```

**Comment by mlh758** [comment]
That helper function looks pretty good. I just have two comments on it:

1. Can you just use `document.body` instead of `document.getElementsByTagName('body')[0]`?
2. It might be easier as two methods. One that gets the pixel value from the dimension text, and one that  gives the min/max determination. Would keep the return value to a simple value, and could avoid the nested ternary.

**Comment by SteveTheTechie** [comment]
Ok, let me tinker a bit more with it. ;-)

**Comment by SteveTheTechie** [comment]
I just discovered that innerWidth() also includes padding, so that must be accounted for, also.

In the following image, the border belongs to a parent div that I have set padding: 10px on.  The native select has a width of 100%.   I have also set the width of the multiselect button at 100%, but now it accounts for the padding in the parent div.

<img width="500" alt="capture-ms-padding" src="https://user-images.githubusercontent.com/5685883/35774385-6979842e-0922-11e8-93b7-dc8dc93d81e7.PNG">

**Comment by SteveTheTechie** [comment]
I now have a working update.  :-)

> **Sample Options:**
>----
> buttonWidth: '20em', 
> menuWidth: null,     
> height: '2in',            

(Menu width gets set to button width for above.)
<img width="214" alt="capture-menuwidth null" src="https://user-images.githubusercontent.com/5685883/35776806-0bb8140a-0958-11e8-9289-e596c6597fa4.PNG">

> **Sample Options:**
>----
> buttonWidth: '20em', 
> menuWidth: 'auto',     
> height: '2in',            
<img width="218" alt="capture-menuwidth auto" src="https://user-images.githubusercontent.com/5685883/35776829-5eacad10-0958-11e8-9664-9c012e089e62.PNG">

**Comment by SteveTheTechie** [comment]
The following is worth taking a look at for your awareness, if you have some time:
https://github.com/select2/select2/issues?utf8=%E2%9C%93&q=is%3Aissue+getBoundingClientRect

These reference a floating point error issue in the jQuery width functions that I also ran into and had to correct for.

**Comment by mlh758** [comment]
One thing I just noticed looking back through your previous comments:

>Unless a max-width is set to cause wrapping at some point (w/ enough list selections shown), the button will grow in width forever. (?) Therefore, an implicit max-width equal to window.innerWidth would be wise--this would cause the displayed list selections to wrap at that point.

Wouldn't that be where the `# of # selected` or passing a function in for that display would be useful?

It looks like they weren't using `getBoundingClientRect` due to IE8 support. I've already stopped supporting anything older than IE10 starting with version 2.x however, so if that makes your life easier you could just use that outright. I work in the healthcare sector and even we stopped supporting IE < 10...

**Comment by SteveTheTechie** [comment]
>Wouldn't that be where the # of # selected or passing a function in for that display would be useful?

Yes... letting it grow forever would be an edge case.  (I think I was tired when I wrote that.)

I got getBoundingClientRect handled w/o any problems.  If you calc the widths from .right - .left, that is supported back to IE5.  😏  no worries.   4 new helper functions:  _getBCRWidth(elem), _getBCRHeight(elem), _jqWidthFix(elem), and _jqHeightFix(elem).   The latter 2 are for cases when you need something other than outerWidth(false) or outerHeight(false) which is what the "BCR Width" and "BCR Height" correspond to.  (I think I am only using them a couple times each.)

With all the changes, I have 9 unit tests failing...  I was expecting worse.  I will fix those and post another PR.  Hopefully I will not mangle the white space too much this time.  😆

**Comment by mlh758** [comment]
> Hopefully I will not mangle the white space too much this time. 😉

Hah, better than mangling the css like I did.

**Comment by SteveTheTechie** [comment]
Using updated tests with updated Qunit....
<img width="321" alt="capture-tests-passed" src="https://user-images.githubusercontent.com/5685883/35837248-a7da2a34-0a99-11e8-9f2b-4755fe22e016.PNG">

**Comment by SteveTheTechie** [comment]
Just a few closing comments for "posterity" on why I implemented the update the way I did.

1.  `minWidth` was always a confusing option to me.   It was never clear to me what part of the width you were specifying a minimum width for.  Because of that, I chose to get rid of it and go with a more explicit approach that allowed specifying a minimum with for either the button or the menu (or both, really).

2.  The `_getMinWidth()` function was almost a unit converter in that it converted % widths to pixel widths.  I thought that unit conversions may be helpful, so I chose to keep and expand on that in a simple helper function, `_parse2px()` that is used in several places.

3.  For most use cases, I tend to think that an explicit height of the menu should not need to be specified.  If the size attribute is used on the underlying select, that really can implicitly set the height.  Otherwise, for most cases, you should just be able to have the widget show all the items (auto height) up the limit of the available height of the window.

4.  Reference: https://stackoverflow.com/questions/450903/how-to-make-div-not-larger-than-its-contents  
I am aware that using floats for determination of intrinsic ("shink-wrapped") widths of the menu header and checkboxes could potentially be problematic in some edge use cases.  I went with them because they are simple to apply and remove for measurements via a class and keep the layout change minimal. However, I did consider the alternatives (keep in mind that these are all temporarily applied just to get a width meaurement):
 - It became clear that I did not want to use a clone of the menu for this.  It is too slow and creating a lot of elements strictly for the purpose of measuring seems a bit silly.
 - Using `display: inline-block` would cause the header and the check boxes to align side by side with a separator margin.  I thought that may cause flashes from the layout changing, so I decided against that.
-  Using `position: absolute` could be used, but I would have save the current position setting and restore it.  (a minor nit pick)  If floats are decided against in the future, I would probably use this approach instead.
- I tried `display: table`, but it did not observe overflow setting of the parent menu container, so I decided against it.

5.  `_getScrollBarWidth()` uses a "brute-force" addition of a div and measuring it as a fall-back for lack of $.position support.  A more elegant fall-back of determining the scroll bar width should maybe be implemented in the future.

**Comment by SteveTheTechie** [comment]
@mlh758:  As I believe we are in agreement that the related PR is ready to be merged, I am closing this issue.  I am pretty happy this is now in the "rear view mirror".  😃
